rosbag2_storage/src/rosbag2_storage/qos.cpp

```diff
@@ -218,10 +359,51 @@ Rosbag2QoS Rosbag2QoS::adapt_offer_to_recorded_offers(
     return result.default_history();
   }
 
-  ROSBAG2_TRANSPORT_LOG_WARN_STREAM(
+  ROSBAG2_STORAGE_LOG_WARN_STREAM(
     "Not all original publishers on topic " << topic_name << " offered the same QoS profiles. "
       "Rosbag2 cannot yet choose an adapted profile to offer for this mixed case. "
-      "Falling back to the rosbag2_transport default publisher offer.");
+      "Falling back to the rosbag2_storage default publisher offer.");
   return Rosbag2QoS{};
 }
-}  // namespace rosbag2_transport
+
+std::vector<rosbag2_storage::Rosbag2QoS> from_rclcpp_qos_vector(std::vector<rclcpp::QoS> in)
+{
+  std::vector<rosbag2_storage::Rosbag2QoS> out;
+  out.reserve(in.size());
+  std::transform(
+    in.begin(), in.end(), std::back_inserter(out),
+    [](auto & qos) {return static_cast<rosbag2_storage::Rosbag2QoS>(qos);});
+  return out;
+}
+
+std::string serialize_rclcpp_qos_vector(std::vector<rclcpp::QoS> in)
+{
+  std::vector<rosbag2_storage::Rosbag2QoS> to_encode = from_rclcpp_qos_vector(in);
+  auto node = YAML::convert<std::vector<rosbag2_storage::Rosbag2QoS>>::encode(to_encode);
+  return YAML::Dump(node);
```

Comment:
Curious if can eliminate extra copy `to_encode` by providing the direct convert function for std::vector<rclcpp::QoS> ?

Reply:
I think I found a simple way to simplify a lot here.

`YAML::convert<std::vector<rclcpp::QoS>>` can be implemented (as the std::vector is default constructible).